### PR TITLE
Refactor schema for storing selection results

### PR DESCRIPTION
**Pull request by philiptzou**
The database schema for selection results (in-vivo and in-vitro) is directly originated from early CSV files. The current tables include all needed columns but 1) lack of normalization, 2) not following the established naming convention, and 3) not constrained by existing tables. The purpose of this pull request is to redesign the selection result tables and their derived tables.

Note: the code is still under construction. Please feel free to comment with your thoughts :)

### CODE REVIEW

File: schema.dbml [hunk 1]
```diff
@@ -600,39 +600,30 @@ Table invitro_selection_results as Vitro {
...
-  // TODO: refactor to use subjects table
+Enum antibody_dosage_unit_enum {
+  "mg/kg"
+}
+
+Table rx_invivo as RxVivo {
   ref_name varchar
   rx_name varchar
-  dosage varchar
-  host varchar
```
Comment (philiptzou) [inline schema.dbml:614]: `host` now becomes `subjects.species`.

File: schema.dbml [hunk 2]
```diff
@@ -600,39 +600,30 @@ Table invitro_selection_results as Vitro {
...
-  infection_name varchar [ref: > I.iso_name]
```
Comment (philiptzou) [inline schema.dbml:615]: `infection_name` now becomes `infected_var_name`, which is a foreign key to `variants.var_name`.

File: schema.dbml [hunk 3]
```diff
@@ -600,39 +600,30 @@ Table invitro_selection_results as Vitro {
...
-  gene varchar [
```
Comment (philiptzou) [inline schema.dbml:616]: `gene`, `position` and `amino_acid` should be stored in `isolate_mutations`, which can be link to this table through `subject_history`.

File: schema.dbml [hunk 4]
```diff
@@ -600,39 +600,30 @@ Table invitro_selection_results as Vitro {
...
-    note: 'Mutation gene'
-  ]
-  position int [
-    note: 'Position of this mutation'
-  ]
-  amino_acid amino_acid_enum [
-    note: 'Amino acid of this mutation'
-  ]
-  subject varchar
```
Comment (philiptzou) [inline schema.dbml:625]: `subject` is renamed to `subject_name` and becomes a foreign key of `subjects.subject_name`. If `num_subjects` is greater than `num_subjects_with_mut`, two subjects should be created and linked to different isolates (with / without mutations).

File: schema.dbml [hunk 5]
```diff
@@ -600,39 +600,30 @@ Table invitro_selection_results as Vitro {
...
-  sampling varchar
-  num_subjects int
```
Comment (philiptzou) [inline schema.dbml:627]: `num_subjects` is removed from the core table. It now becomes a derived value from `subjects.num_subjects`.

File: schema.dbml [hunk 6]
```diff
@@ -600,39 +600,30 @@ Table invitro_selection_results as Vitro {
...
-  num_subjects_with_mut int
```
Comment (philiptzou) [inline schema.dbml:628]: `num_subjects_with_mut` is removed from the core table. The value is now saved at `subjects.num_subjects`.

File: schema.dbml [hunk 7]
```diff
@@ -600,39 +600,30 @@ Table invitro_selection_results as Vitro {
...
+  dosage numeric(2,0)
```
Comment (philiptzou) [inline schema.dbml]: The old "dosage" field is divided to two fields: one to store the number and one to store the unit. We may add a cmp field in the future when it's needed.

File: schema.dbml [hunk 8]
```diff
@@ -600,39 +600,30 @@ Table invitro_selection_results as Vitro {
...
+  dosage_unit antibody_dosage_unit_enum
+  subject_name varchar [ref: > Pt.subject_name]
+  collection_date date
```
Comment (philiptzou) [inline schema.dbml:655]: A "collection_date" is added to be joined with `subject_history.event_date`.

```sql
SELECT iso_name as infected_iso_name
  FROM rx_invivo, subject_history
  WHERE
    rx_invivo.subject_name = subject_history.subject_name AND
    subject_history.event = 'isolation' AND
    rx_invivo.collection_date = subject_history.event_date
```

File: schema.dbml [hunk 9]
```diff
@@ -990,3 +981,33 @@ Table susc_results as Susc {
     (rx_name, control_iso_name, iso_name, ref_name, potency_type, assay_name) [pk]
   }
 }
+
+Table invivo_selection_results {
```
Comment (philiptzou) [inline schema.dbml:1027]: The new `invivo_selection_results` becomes the derived table using the data from `rx_invivo`, `subjects`, `subject_history` and `variant_mutations`. I tried to preserve as many old fields as possible but following our established naming convention.

File: schema.dbml [hunk 10]
```diff
@@ -153,8 +153,13 @@ Enum event_enum {
   "3rd dose isolation"
   "infection"
   "isolation"
+  "culture"
```
Comment (KaimingTao) [inline schema.dbml:156]: what does "culture" meaning? Is that a verb?

File: schema.dbml (hunk 9 above)
Comment (KaimingTao) [reply schema.dbml:1027]: what is "variant_mutations"? How the *position* and *amino_acid* field in *invivo_selection_results* generated by other tables? What if the paper doesn't provide isolates from GISAID by only single point mutations, how to enter the data?

File: schema.dbml (hunk 10 above)
Comment (philiptzou) [reply schema.dbml:156]: It's isolation culture, a noun. Also I plan to refactor the subject_history and remove the the event_enum. We can have subject_vaccinations (1st/2nd/3rd dose), subject_isolations (isolation, culture and 1st/2nd/3rd isolation), subject_infections (infection) and subject_treatments (new, already created).

Comment (KaimingTao) [reply schema.dbml:156]: I'm thinking if a subject get infected and get vaccinated that will need to fill in all of _subject_vaccinations_ and _subject_isolations_, and _subject_infections_.

Comment (philiptzou) [reply schema.dbml:156]: Yes, that would be the case.

File: schema.dbml (hunk 9 above)
Comment (philiptzou) [reply schema.dbml:1027]: Sorry but it's a typo in my comment. It should be `isolate_mutations`. If the paper doesn't provide the whole isolation sequence but just mutations, we just link to the isolate of single mutation/combination mutations.

**Pull request merged**